sys/evtimer/evtimer.c

```diff
@@ -0,0 +1,185 @@
+        if (list_entry == event) {
+            list->next = event->next;
+            if (list->next) {
+                list_entry = list->next;
+                list_entry->offset += event->offset;
+            }
+            break;
+        }
+        list = list->next;
+    }
+}
+
+static void _set_timer(xtimer_t *timer, uint32_t offset)
+{
+    uint64_t offset_in_us = (uint64_t)offset * 1000;
+    DEBUG("evtimer: now=%"PRIu32" setting xtimer to %"PRIu32":%"PRIu32"\n",  xtimer_now(),
+            (uint32_t)(offset_in_us>>32), (uint32_t)(offset_in_us));
+    _xtimer_set64(timer, offset_in_us, offset_in_us>>32);
+}
+
+static void _update_timer(evtimer_t *evtimer)
+{
+    if (evtimer->events) {
+        evtimer_event_t *event = evtimer->events;
+        _set_timer(&evtimer->timer, event->offset);
+    }
+    else {
+        xtimer_remove(&evtimer->timer);
+    }
+}
+
+static uint32_t _get_offset(xtimer_t *timer)
+{
+    uint64_t now = xtimer_now64();
+    uint64_t target = ((uint64_t)timer->long_target)<<32 | timer->target;
+
+    if (target <= now) {
+        return 0;
+    }
+    else {
+        target -= now;
+        /* add half of 125 so integer division rounds to nearest */
+        return div_u64_by_125((target >> 3) + 62);
+    }
+}
+
+static void _update_head_offset(evtimer_t *evtimer)
+{
+    if (evtimer->events) {
+        evtimer_event_t *event = evtimer->events;
+        event->offset = _get_offset(&evtimer->timer);
+        DEBUG("evtimer: _update_head_offset(): new head offset %"PRIu32"\n", event->offset);
+    }
+}
+
+void evtimer_add(evtimer_t *evtimer, evtimer_event_t *event)
+{
+    unsigned state = irq_disable();
+    DEBUG("evtimer_add(): adding event with offset %"PRIu32"\n", event->offset);
+
+    _update_head_offset(evtimer);
+    _add_event_to_list((evtimer_event_t*)&evtimer->events, event);
+
+    if (evtimer->events == event) {
+        _set_timer(&evtimer->timer, event->offset);
+    }
+    irq_restore(state);
+}
+
+void evtimer_del(evtimer_t *evtimer, evtimer_event_t *event)
+{
+    unsigned state = irq_disable();
+    DEBUG("evtimer_del(): removing event with offset %"PRIu32"\n", event->offset);
+    _update_head_offset(evtimer);
+    _del_event_from_list((evtimer_event_t*)&evtimer->events, event);
+    _update_timer(evtimer);
+    irq_restore(state);
+}
+
+static evtimer_event_t *_get_next(evtimer_t *evtimer)
+{
+    evtimer_event_t *event = evtimer->events;
+    if (event && (event->offset == 0)) {
+        evtimer->events = event->next;
+        return event;
+    }
+    else {
+        return NULL;
+    }
+}
+
+void evtimer_msg_handler(void *arg)
+{
+    DEBUG("evtimer_msg_handler()\n");
+
+    evtimer_t *evtimer = (evtimer_t*) arg;
+
+    evtimer_event_t *event = evtimer->events;
+    event->offset = 0;
+
+    evtimer_msg_event_t *mevent;
+    while ((mevent = (evtimer_msg_event_t *)_get_next(evtimer))) {
+        msg_send_int(&mevent->msg, mevent->msg.sender_pid);
```

Comment:
Send to `sender_pid`? Also: isn't this value usually uninitialized by the user?

Easy fix: see [here](https://github.com/kaspar030/RIOT/pull/19/files#diff-86e356dc5223dce7b4926e951244cfc0R77)